pkg/server/apiextensions.go

```diff
@@ -31,137 +31,232 @@ import (
+	}
+
+	return ret, nil
+}
+
+func (p *systemCRDProvider) Keys(ctx context.Context, org, workspace string) sets.String {
+	switch {
+	case workspace == helper.RootCluster:
+		return p.rootCRDs
+	case org == "system":
+		// fall through
+	case org != "":
+		workspaceKey := helper.WorkspaceKey(org, workspace)
+		clusterWorkspace, err := p.getClusterWorkspace(ctx, workspaceKey)
+		if err != nil {
+			// this shouldn't happen. The getters use quorum-read client on cache-miss.
+			klog.Errorf("error getting cluster workspace %q: %v", workspaceKey, err)
+			break
+		}
+		switch clusterWorkspace.Spec.Type {
+		case "Universal":
+			return p.universalCRDs
+		case "Organization":
+			return p.orgCRDs
+		}
+	}
+
+	return sets.NewString()
+}
+
+// apiBindingAwareCRDLister is a CRD lister combines APIs coming from APIBindings with CRDs in a workspace.
+type apiBindingAwareCRDLister struct {
+	kcpClusterClient  kcpclientset.ClusterInterface
+	crdLister         apiextensionslisters.CustomResourceDefinitionLister
+	workspaceLister   tenancylisters.ClusterWorkspaceLister
+	apiBindingLister  apislisters.APIBindingLister
+	systemCRDProvider *systemCRDProvider
+}
+
+var _ apiextensionslisters.CustomResourceDefinitionLister = (*apiBindingAwareCRDLister)(nil)
 
 // List lists all CustomResourceDefinitions in the underlying store matching selector. This method does not
-// support scoping to logical clusters or workspace inheritance.
-func (c *inheritanceCRDLister) List(selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+// support scoping to logical clusters or APIBindings.
+func (c *apiBindingAwareCRDLister) List(selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
 	return c.crdLister.ListWithContext(context.Background(), selector)
 }
 
-// ListWithContext lists all CustomResourceDefinitions in the logical cluster associated with ctx that match
-// selector. ClusterWorkspace API inheritance is also supported: if the ClusterWorkspace for ctx's logical cluster
-// has spec.inheritFrom set, it will aggregate all CustomResourceDefinitions from the ClusterWorkspace named
-// spec.inheritFrom with the CustomResourceDefinitions from the ClusterWorkspace for ctx's logical cluster.
-func (c *inheritanceCRDLister) ListWithContext(ctx context.Context, selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+// ListWithContext lists all CustomResourceDefinitions that come in via APIBindings as well as all in the current
+// logical cluster.
+func (c *apiBindingAwareCRDLister) ListWithContext(ctx context.Context, selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
 	cluster, err := genericapirequest.ValidClusterFrom(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	// Check for API inheritance
-	inheriting := false
-	inheritFrom := ""
-	orgName, ws, err := helper.ParseLogicalClusterName(cluster.Name)
+	var ret []*apiextensionsv1.CustomResourceDefinition
```

Comment:
(No action required) Maybe define where it is first used (e.g. L180)?

Reply:
done